apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/AgentOverheadMetrics.java

```diff
@@ -0,0 +1,442 @@
+            return false;
+        }
+        try {
+            boolean isSupported = (Boolean) isThreadAllocatedMemorySupported.invoke(threadBean);
+            if (!isSupported) {
+                logger.warn("Agent allocation metrics can not be enabled: ThreadMxBean.isThreadAllocatedMemorySupported() returned false");
+                return false;
+            }
+        } catch (Exception e) {
+            logger.warn("Agent allocation metrics can not be enabled", e);
+            return false;
+        }
+        try {
+            setThreadAllocatedMemoryEnabled.invoke(threadBean, true);
+            logger.debug("Enabled agent allocation measurement");
+        } catch (Exception e) {
+            logger.warn("Agent allocation metrics can not be enabled", e);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean enableThreadCpuTimeMeasurement() {
+        boolean isSupported = threadBean.isThreadCpuTimeSupported();
+        if (!isSupported) {
+            logger.warn("Agent cpu overhead metrics can not be enabled: ThreadMxBean.isThreadCpuTimeSupported() returned false");
+            return false;
+        }
+        threadBean.setThreadCpuTimeEnabled(true);
+        return true;
+    }
+
+    private boolean anyMetricEnabled() {
+        return cpuOverheadMetricEnabled || cpuUsageMetricEnabled || allocationMetricEnabled || threadCountMetricEnabled;
+    }
+
+    @Override
+    public void elasticThreadStarted(Thread thread, String purpose) {
+        lastThreadInfo.putIfAbsent(thread, new ThreadInfo(purpose));
+        ThreadInfo threadInfo = lastThreadInfo.get(thread);
+        updateAllocationStat(thread, threadInfo);
+        updateCpuTimeStat(thread, threadInfo);
+    }
+
+    @Override
+    public void elasticThreadFinished(Thread thread) {
+        ThreadInfo threadInfo = lastThreadInfo.get(thread);
+        if (threadInfo != null) {
+            if (cpuOverheadMetricEnabled || cpuUsageMetricEnabled) {
+                threadInfo.deathCpuTime = getThreadCpuTime(thread);
+            }
+            if (allocationMetricEnabled) {
+                threadInfo.deathAllocationBytes = getThreadAllocatedBytes(thread);
+            }
+        }
+    }
+
+    @Override
+    public void collectAndReset(MetricCollector collector) {
+        collectCpuUsageMetrics(collector);
+        collectAllocationMetrics(collector);
+        collectActiveThreadsMetric(collector);
+
+        //cleanup dead threads
+        for (Map.Entry<Thread, ThreadInfo> threadInfo : lastThreadInfo.entrySet()) {
+            Thread thread = threadInfo.getKey();
+            if (!thread.isAlive()) {
+                lastThreadInfo.remove(thread);
+            }
+        }
+    }
+
+    private void collectActiveThreadsMetric(MetricCollector collector) {
+        if (!threadCountMetricEnabled) {
+            return;
+        }
+        Map<String, AtomicLong> threadCountByPurpose = new HashMap<>();
+
```

Comment:
can we reuse the Map at least for these 3 calls, so one map created and dumped instead of 3. I don't think it's necessary to reuse it across different collectAndReset calls since that is multiple seconds apart and one map per call is acceptable. 3 maps probably are too, but it's so easy to reuse that one we might as well cut down that garbage